text/0000-api-evolution.md

```diff
@@ -0,0 +1,748 @@
+
+```rust
+// MAJOR CHANGE
+
+// Before
+struct Foo<T = u8>(pub T, pub u8);
+
+// After
+struct Foo<T = u8>(pub T, pub T);
+```
+
+since there may be existing uses of `Foo` with a non-default type parameter
+which would break as a result of the change.
+
+It's also permitted to change from a generic type to a more-generic one in a
+minor revision:
+
+```rust
+// MINOR CHANGE
+
+// Before
+struct Foo<T>(pub T, pub T);
+
+// After
+struct Foo<T, U = T>(pub T, pub U);
+```
+
+since, again, all existing uses of the type `Foo<T>` will yield the same field
+types as before.
+
+### Signatures in functions
+
+All of the changes mentioned below are considered major changes in the context
+of trait methods, since they can break implementors.
+
+#### Major change: adding new arguments.
+
+At the moment, Rust does not provide defaulted arguments, so any change in arity
+is a breaking change.
+
+#### Minor change: introducing a new type parameter.
+
+Technically, adding a (non-defaulted) type parameter can break code:
+
+```rust
+// MINOR CHANGE (but causes breakage)
+
+// Before
+fn foo<T>(...) { ... }
+
+// After
+fn foo<T, U>(...) { ... }
+```
+
+will break any calls like `foo::<u8>`. However, such explicit calls are rare
+enough (and can usually be written in other ways) that this breakage is
+considered minor. (However, one should take into account how likely it is that
+the function in question is being called with explicit type arguments).
+
+Such changes are an important ingredient of abstracting to use generics, as described next.
+
+#### Minor change: generalizing to generics.
+
+The type of an argument to a function, or its return value, can be *generalized*
+to use generics, including by introducing a new type parameter (as long as it
+can be instantiated to the original type). For example, the following change is
+allowed:
+
+```rust
+// MINOR CHANGE
+
+// Before
+fn foo(x: u8) -> u8;
+fn bar<T: Iterator<Item = u8>>(t: T);
+
+// After
+fn foo<T: Add>(x: T) -> T;
+fn bar<T: IntoIterator<Item = u8>>(t: T);
+```
+
+because all existing uses are instantiations of the new signature. On the other
+hand, the following isn't allowed in a minor revision:
+
+```rust
+// MAJOR CHANGE
+
+// Before
+fn foo(x: Vec<u8>);
+
+// After
+fn foo<T: Copy + IntoIterator<Item = u8>>(x: T);
+```
+
+because the generics include a constraint not satisfied by the original type.
+
+Introducing generics in this way can potentially create type inference failures,
+but these are considered acceptable per the principles of the RFC: they only
+require local annotations that could have been inserted in advance.
+
+Perhaps somewhat surprisingly, generalization applies to trait objects as well,
+given that every trait implements itself:
+
+```rust
+// MINOR CHANGE
+
+// Before
+fn foo(t: &Trait);
+
+// After
+fn foo<T: Trait + ?Sized>(t: &T);
+```
+
+(The use of `?Sized` is essential; otherwise you couldn't recover the original
+signature).
```

Comment:
@theemathas: You can not create trait objects for unsized types, so if eg `str` implemented `Trait` you could put it in the "before" case, but not in the "after" case.